cdc/scheduler/internal/v3/keyspan/reconciler.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keyspan
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/processor/tablepb"
+	"github.com/pingcap/tiflow/cdc/scheduler/internal/v3/compat"
+	"github.com/pingcap/tiflow/cdc/scheduler/internal/v3/replication"
+	"github.com/pingcap/tiflow/pkg/config"
+	"github.com/pingcap/tiflow/pkg/spanz"
+	"github.com/tikv/client-go/v2/tikv"
+	"go.uber.org/zap"
+)
+
+// Reconciler reconciles span and table mapping, make sure spans are in
+// a desired state and covers all table ranges.
+type Reconciler struct {
+	tableSpans map[model.TableID][]tablepb.Span
+	spanCache  []tablepb.Span
+
+	regionCache      RegionCache
+	changefeedID     model.ChangeFeedID
+	maxRegionPerSpan int
+}
+
+// NewReconciler returns a Reconciler.
+func NewReconciler(
+	changefeedID model.ChangeFeedID, regionCache RegionCache,
+	cfg *config.SchedulerConfig,
+) *Reconciler {
+	return &Reconciler{
+		tableSpans:       make(map[int64][]tablepb.Span),
+		regionCache:      regionCache,
+		maxRegionPerSpan: cfg.RegionPerSpan,
+	}
+}
+
+// Reconcile spans that need to be replicated based on current cluster status.
+// It handles following cases:
+// 1. Changefeed initialization
+// 2. owner switch.
+// 3. Owner switch after some captures fail.
+// 4. Add table by DDL.
+// 5. Drop table by DDL.
+// 6. Some captures fail, does NOT affect spans.
+func (m *Reconciler) Reconcile(
```

Comment:
The method looks quite complicated, is it possible to split it into multiple small methods?

Reply:
It's in performance critical path, splitting to multiple methods may lower its speed.
See https://gist.github.com/overvenus/475b333c7726a97d051e01484deb0bf0